flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkImageBuilder.java

```diff
@@ -0,0 +1,240 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.tests.util.flink.container;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.GlobalConfiguration;
+import org.apache.flink.tests.util.util.FileUtils;
+
+import com.github.dockerjava.api.exception.NotFoundException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A builder class for constructing Docker image based on flink-dist. */
+public class FlinkImageBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkImageBuilder.class);
+    private static final String FLINK_BASE_IMAGE_NAME = "flink-dist-base";
+    private static final String DEFAULT_IMAGE_NAME = "flink-dist-configured";
+    private static final Duration DEFAULT_TIMEOUT = Duration.ofMinutes(5);
+
+    private final Map<Path, Path> filesToCopy = new HashMap<>();
+
+    private String imageName = DEFAULT_IMAGE_NAME;
+    private String imageNameSuffix;
+    private Path flinkDist = FileUtils.findFlinkDist();
+    private String javaVersion;
+    private Configuration conf;
+    private Duration timeout = DEFAULT_TIMEOUT;
+    private String startupCommand;
+
+    /**
+     * Sets the name of building image.
+     *
+     * <p>If the name is not specified, {@link #DEFAULT_IMAGE_NAME} will be used.
+     */
+    public FlinkImageBuilder setImageName(String imageName) {
+        this.imageName = imageName;
+        return this;
+    }
+
+    /**
+     * Sets the path of flink-dist directory.
+     *
+     * <p>If path is not specified, the dist directory under current project will be used.
+     */
+    public FlinkImageBuilder setFlinkDistPath(Path flinkDist) {
+        this.flinkDist = flinkDist;
+        return this;
+    }
+
+    /**
+     * Sets JDK version in the image.
+     *
+     * <p>If version is not specified, the JDK version of the current JVM will be used.
```

Comment:
Nit: you can add a link to the docker hub of openjdk because the version format is the tag of the OpenJDK image.